### PR TITLE
Users/arjgupta/azurePsSxS

### PR DESCRIPTION
**Pull request by arjgupta**

### CODE REVIEW

File: Tasks/Common/VstsAzureHelpers_/ImportFunctions.ps1 [hunk 1]
```diff
@@ -3,35 +3,48 @@
     param(
         [Parameter(Mandatory = $true)]
         [ValidateSet('Azure', 'AzureRM')]
-        [string[]]$PreferredModule)
+        [string[]]$PreferredModule,
```
Comment (kmkumaran) [inline Tasks/Common/VstsAzureHelpers_/ImportFunctions.ps1]: Keep the formatting consistent for all arguments. Have space between the type and variable. I see this in other functions as well. Please correct in all instances.

File: Tasks/Common/VstsAzureHelpers_/ImportFunctions.ps1 [hunk 2]
```diff
@@ -115,18 +137,21 @@ function Import-FromSdkPath {
             $path = [System.IO.Path]::Combine($programFiles, $partialPath)
             Write-Verbose "Checking if path exists: $path"
             if (Test-Path -LiteralPath $path -PathType Leaf) {
+                if ($azurePsVersion -and (-not ($(Get-SdkVersion) -eq $azurePsVersion))) {
+                    continue
```
Comment (kmkumaran) [inline Tasks/Common/VstsAzureHelpers_/ImportFunctions.ps1]: Why should we continue? There can be only one SDK version installed in a box. Given that it does not match why should we continue?

File: Tasks/Common/VstsAzureHelpers_/ImportFunctions.ps1 [hunk 3]
```diff
@@ -136,3 +161,49 @@ function Import-FromSdkPath {
         Trace-VstsLeavingInvocation $MyInvocation
     }
 }
+
+function Get-SdkVersion {
+    Trace-VstsEnteringInvocation $MyInvocation
+    try{
+        $regKey = "HKLM:\Software\Wow6432Node\Microsoft\Windows\CurrentVersion\Uninstall\*"
+        $installedApplications = Get-ItemProperty -Path $regKey
+        $SdkVersion = ($installedApplications | Where-Object { $_.DisplayName -and $_.DisplayName.toLower().Contains("microsoft azure powershell") } | Select-Object -First 1).DisplayVersion
```
Comment (arjgupta) [inline Tasks/Common/VstsAzureHelpers_/ImportFunctions.ps1]: yes, display name is not localized

File: Tasks/Common/VstsAzureHelpers_/ImportFunctions.ps1 (hunk 2 above)
Comment (arjgupta) [inline Tasks/Common/VstsAzureHelpers_/ImportFunctions.ps1]: the loop will exit in the next iteration and return false in the end

File: Tasks/Common/VstsAzureHelpers_/ImportFunctions.ps1 [hunk 4]
```diff
@@ -61,9 +75,18 @@ function Import-FromModulePath {
 
         # Attempt to resolve the module.
         Write-Verbose "Attempting to find the module '$name' from the module path."
-        $module = Get-Module -Name $name -ListAvailable | Select-Object -First 1
-        if (!$module) {
-            return $false
+        if ($azurePsVersion) {
+            $module = Get-Module -Name $name -ListAvailable | Where-Object {$_.Version -eq $azurePsVersion} | Select-Object -First 1
+            if (!$module) {
```
Comment (arjgupta) [inline Tasks/Common/VstsAzureHelpers_/ImportFunctions.ps1]: done

File: Tasks/AzurePowerShell/AzurePowerShell.ps1 [hunk 5]
```diff
@@ -19,11 +21,26 @@ if ($scriptArguments -match '[\r\n]') {
     throw (Get-VstsLocString -Key InvalidScriptArguments0 -ArgumentList $scriptArguments)
 }
 
+if($targetAzurePs -eq "OtherVersion") {
```
Comment (chshrikh) [inline Tasks/AzurePowerShell/AzurePowerShell.ps1]: will be better if we can make "OtherVersion""LatestVersion" as constant variables

File: Tasks/AzurePowerShell/AzurePowerShell.ps1 [hunk 6]
```diff
@@ -19,11 +21,26 @@ if ($scriptArguments -match '[\r\n]') {
...
+    $targetAzurePs = $customTargetAzurePs.Trim()
```
Comment (chshrikh) [inline Tasks/AzurePowerShell/AzurePowerShell.ps1]: check this for null and empty before making it trim

File: Tasks/AzurePowerShell/AzurePowerShell.ps1 [hunk 7]
```diff
@@ -6,6 +6,8 @@ $scriptType = Get-VstsInput -Name ScriptType -Require
 $scriptPath = Get-VstsInput -Name ScriptPath
 $scriptInline = Get-VstsInput -Name Inline
 $scriptArguments = Get-VstsInput -Name ScriptArguments
+$targetAzurePs = Get-VstsInput -Name TargetAzurePs
```
Comment (chshrikh) [inline Tasks/AzurePowerShell/AzurePowerShell.ps1]: add logging for targetAzurePs  and customTargetAzurePs  if Get-VstsInput is not doing

File: Tasks/AzurePowerShell/AzurePowerShell.ps1 [hunk 8]
```diff
@@ -19,11 +21,26 @@ if ($scriptArguments -match '[\r\n]') {
     throw (Get-VstsLocString -Key InvalidScriptArguments0 -ArgumentList $scriptArguments)
 }
 
+if($targetAzurePs -eq "OtherVersion") {
+    $targetAzurePs = $customTargetAzurePs.Trim()
+}
+
+$pattern = "^[0-9]+\.[0-9]+\.[0-9]+$"
```
Comment (chshrikh) [inline Tasks/AzurePowerShell/AzurePowerShell.ps1]: can we be explicit in help mark for not supporting regex for version like - 2.3.*, 3.*,.

File: Tasks/AzurePowerShell/Utility.ps1 [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+    if ($targetAzurePs) {
+        $hostedAgentAzureRmModulePath = $env:SystemDrive + "\Modules\AzureRm_" + $targetAzurePs
+        $hostedAgentAzureModulePath = $env:SystemDrive + "\Modules\Azure_" + $targetAzurePs
+    }
+    else {
+        $hostedAgentAzureRmModulePath = Get-LatestModule -patternToMatch "^azurerm_[0-9]+\.[0-9]+\.[0-9]+$" -patternToExtract "[0-9]+\.[0-9]+\.[0-9]+$"
+        $hostedAgentAzureModulePath  =  Get-LatestModule -patternToMatch "^azure_[0-9]+\.[0-9]+\.[0-9]+$"   -patternToExtract "[0-9]+\.[0-9]+\.[0-9]+$"
+    }
+    $env:PSModulePath = $hostedAgentAzureRmModulePath + ";" + $env:PSModulePath
```
Comment (chshrikh) [inline Tasks/AzurePowerShell/Utility.ps1]: adding logs will be better here like updating the module path, updated module path

File: Tasks/Common/VstsAzureHelpers_/ImportFunctions.ps1 [hunk 10]
```diff
@@ -76,17 +100,15 @@ function Import-FromModulePath {
             $script:azureModule = $module
         } else {
             # The AzureRM module was imported.
-
             # Validate the AzureRM.profile module can be found.
-            $profileModule = Get-Module -Name AzureRM.profile -ListAvailable | Select-Object -First 1
+            $profileModule = (Get-Module -Name AzureRM).NestedModules | Where-Object { $_.Name.toLower() -eq "azurerm.profile" }
```
Comment (chshrikh) [inline Tasks/Common/VstsAzureHelpers_/ImportFunctions.ps1]: will suggest to put this in try catch from line 104-111 will help to debug in future

File: Tasks/AzurePowerShell/AzurePowerShell.ps1 (hunk 7 above)
Comment (arjgupta) [inline Tasks/AzurePowerShell/AzurePowerShell.ps1]: all inputs get logged in the debug stream

File: Tasks/AzurePowerShell/AzurePowerShell.ps1 (hunk 5 above)
Comment (arjgupta) [inline Tasks/AzurePowerShell/AzurePowerShell.ps1]: done

File: Tasks/AzurePowerShell/AzurePowerShell.ps1 (hunk 6 above)
Comment (arjgupta) [inline Tasks/AzurePowerShell/AzurePowerShell.ps1]: done

File: Tasks/AzurePowerShell/Utility.ps1 [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    elseif ($targetAzurePs -eq "LatestVersion") {
+        # For Hosted Agent, the Latest Version is 4.1.0
+        $hostedAgentAzureRmModulePath = Get-LatestModule -patternToMatch "^azurerm_[0-9]+\.[0-9]+\.[0-9]+$" -patternToExtract "[0-9]+\.[0-9]+\.[0-9]+$"
+        $hostedAgentAzureModulePath  =  Get-LatestModule -patternToMatch "^azure_[0-9]+\.[0-9]+\.[0-9]+$"   -patternToExtract "[0-9]+\.[0-9]+\.[0-9]+$"
+        $targetAzurePs = ""
+    }
+    else {
+        throw (Get-VstsLocString -Key InvalidVersion -ArgumentList $targetAzurePs)
+    }
+    $env:PSModulePath = $hostedAgentAzureRmModulePath + ";" + $env:PSModulePath
```
Comment (arjgupta) [inline Tasks/AzurePowerShell/Utility.ps1]: done

File: Tasks/AzurePowerShell/Utility.ps1 (hunk 9 above)
Comment (arjgupta) [inline Tasks/AzurePowerShell/Utility.ps1]: done

**Pull request merged**